### PR TITLE
Api validate text to return an array of invalid characters with their names

### PR DESCRIPTION
**Pull request by 70ray**
Following pattern of bad words api.

Sandbox at: https://www.pgdp.org/~rp31/c.branch/api_validate

### CODE REVIEW

File: api/dp-openapi.yaml [hunk 1]
```diff
@@ -963,11 +963,18 @@ paths:
...
         200:
-          description: analysis data
+          description: invalid characters
           content:
             application/json:
               schema:
-                $ref: '#/components/schemas/text_analysis'
+                type: object
+                properties:
+                  invalidChars:
```
Comment (cpeel) [inline api/dp-openapi.yaml:972]:
```suggestion
                  invalid_chars:
```

File: api/v1_projects.inc [hunk 2]
```diff
@@ -793,10 +793,11 @@ function api_v1_project_checkout($method, array $data, array $query_params)
...
 }
 
-function api_v1_project_validatetext($method, $data, array $query_params)
+function api_v1_project_validatetext(string $method, array $data, array $query_params)
 {
-    $text_checker = new TextChecker();
-    return $text_checker->analyse(receive_project_text_from_request_body('text'), $data[":projectid"]);
+    $project = $data[":projectid"];
+    $invalid_characters = $project->find_invalid_characters(receive_project_text_from_request_body('text'));
+    return ["invalidChars" => $invalid_characters];
```
Comment (cpeel) [inline api/v1_projects.inc:800]:
```suggestion
    return ["invalid_chars" => $invalid_characters];
```

**Review by cpeel — COMMENTED** [review]
<nod> This makes more sense to me too, thanks for making the change.

We should update the description on `api/dp-openapi.yaml` line 959:
```
description: Checks the given text for invalid characters and returns any found
```

**Comment by cpeel** [comment]
Missing this update to the description:

> We should update the description on `api/dp-openapi.yaml` line 959:
> 
> ```
> description: Checks the given text for invalid characters and returns any found
> ```

**Comment by 70ray** [comment]
> and returns any found

Done this now.

File: SETUP/tests/unittests/ApiTest.php [hunk 3]
```diff
@@ -678,17 +678,7 @@ public function test_validate_text()
...
-            'mark_array' => [["This is an inv", "", 0], ["Ā", "LATIN CAPITAL LETTER A WITH MACRON", 1], ["lid test file", "", 0]],
-        ];
-        $this->assertEquals($expected, $response);
-
-        $response = $this->validate_text($project->projectid, "This is a valid test file");
-        $expected = [
-            'valid' => true,
-            'mark_array' => [["This is a valid test file", "", 0]],
-        ];
+        $expected = ["invalid_chars" => ["Ā" => "LATIN CAPITAL LETTER A WITH MACRON"]];
```
[the remark this answers is not in the document]
Comment (70ray) [reply SETUP/tests/unittests/ApiTest.php:681]: Yes, why not?

**Review by cpeel — APPROVED** [review]

**Review by jchaffraix — APPROVED** [review]

**Review by srjfoo — APPROVED** [review]

**Pull request merged**